File: api/src/Entity/Utilisateur.php
<?php

namespace App\Entity;

use App\Repository\UtilisateurRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface;
use Symfony\Component\Security\Core\User\UserInterface;

#[ORM\Entity(repositoryClass: UtilisateurRepository::class)]

class Utilisateur implements UserInterface, PasswordAuthenticatedUserInterface
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(length: 180, unique: true)]
    private ?string $email = null;

    #[ORM\Column]
    private ?string $nom = null;

    #[ORM\Column]
    private ?string $prenom = null;


    #[ORM\Column]
    private array $roles = [];

    #[ORM\Column]
    private ?string $password = null;

    #[ORM\Column]
    private ?int $code = null;

    #[ORM\Column]
    private ?bool $verif = false;

    #[ORM\OneToMany(mappedBy: 'proprietaire', targetEntity: Ressource::class)]
    private Collection $proprietaireRessource;

    #[ORM\OneToMany(mappedBy: 'Utilisateur', targetEntity: Commentaire::class)]
    private Collection $commentaires;

    #[ORM\OneToMany(mappedBy: 'Utilisateur', targetEntity: Progression::class)]
    private Collection $progressions;

    #[ORM\ManyToMany(targetEntity: Ressource::class, mappedBy: 'voirRessource')]
    private Collection $ressources;

    public function __construct()
    {
        $this->proprietaireRessource = new ArrayCollection();
        $this->commentaires = new ArrayCollection();
        $this->progressions = new ArrayCollection();
        $this->ressources = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getEmail(): ?string
    {
        return $this->email;
    }

    public function setEmail(string $email): static
    {
        $this->email = $email;

        return $this;
    }

    public function getNom(): ?string
    {
        return $this->nom;
    }

    public function setNom(?string $nom): void
    {
        $this->nom = $nom;
    }

    public function getPrenom(): ?string
    {
        return $this->prenom;
    }

    public function setPrenom(?string $prenom): void
    {
        $this->prenom = $prenom;
    }



    /**
     * A visual identifier that represents this user.
     *
     * @see UserInterface
     */
    public function getUserIdentifier(): string
    {
        return (string) $this->email;
    }

    /**
     * @see UserInterface
     */
    public function getRoles(): array
    {
        $roles = $this->roles;
        // guarantee every user at least has ROLE_USER
        $roles[] = 'ROLE_USER';

        return array_unique($roles);
    }

    public function setRoles(array $roles): static
    {
        $this->roles = $roles;

        return $this;
    }

    public function getPassword(): string
    {
        return $this->password ?? '';
    }

    public function setPassword(string $password): static
    {
        $this->password = $password;

        return $this;
    }

    public function getCode(): ?int
    {
        return $this->code;
    }

    public function setCode(?int $code): void
    {
        $this->code = $code;
    }

    public function getVerif(): ?bool
    {
        return $this->verif;
    }

    public function setVerif(?bool $verif): void
    {
        $this->verif = $verif;
    }

    /**
     * @see UserInterface
     */
    public function eraseCredentials(): void
    {
        // If you store any temporary, sensitive data on the user, clear it here
        // $this->plainPassword = null;
    }

    /**
     * @return Collection<int, Ressource>
     */
    public function getProprietaireRessource(): Collection
    {
        return $this->proprietaireRessource;
    }

    public function addProprietaireRessource(Ressource $proprietaireRessource): static
    {
        if (!$this->proprietaireRessource->contains($proprietaireRessource)) {
            $this->proprietaireRessource->add($proprietaireRessource);
            $proprietaireRessource->setProprietaire($this);
        }

        return $this;
    }

    public function removeProprietaireRessource(Ressource $proprietaireRessource): static
    {
        if ($this->proprietaireRessource->removeElement($proprietaireRessource)) {
            // set the owning side to null (unless already changed)
            if ($proprietaireRessource->getProprietaire() === $this) {
                $proprietaireRessource->setProprietaire(null);
            }
        }

        return $this;
    }

    /**
     * @return Collection<int, Commentaire>
     */
    public function getCommentaires(): Collection
    {
        return $this->commentaires;
    }

    public function addCommentaire(Commentaire $commentaire): static
    {
        if (!$this->commentaires->contains($commentaire)) {
            $this->commentaires->add($commentaire);
            $commentaire->setUtilisateur($this);
        }

        return $this;
    }

    public function removeCommentaire(Commentaire $commentaire): static
    {
        if ($this->commentaires->removeElement($commentaire)) {
            // set the owning side to null (unless already changed)
            if ($commentaire->getUtilisateur() === $this) {
                $commentaire->setUtilisateur(null);
            }
        }

        return $this;
    }

    /**
     * @return Collection<int, Progression>
     */
    public function getProgressions(): Collection
    {
        return $this->progressions;
    }

    public function addProgression(Progression $progression): static
    {
        if (!$this->progressions->contains($progression)) {
            $this->progressions->add($progression);
            $progression->setUtilisateur($this);
        }

        return $this;
    }

    public function removeProgression(Progression $progression): static
    {
        if ($this->progressions->removeElement($progression)) {
            // set the owning side to null (unless already changed)
            if ($progression->getUtilisateur() === $this) {
                $progression->setUtilisateur(null);
            }
        }

        return $this;
    }

    public function toArray(): array
    {
        return [
            'id' => $this->id,
            'email' => $this->email,
            'nom' => $this->nom,
            'prenom' => $this->prenom,
            'roles' => $this->roles,
            'code' => $this->code,
            'verif' => $this->verif,
        ];
    }

    /**
     * @return Collection<int, Ressource>
     */
    public function getRessources(): Collection
    {
        return $this->ressources;
    }

    public function addRessource(Ressource $ressource): static
    {
        if (!$this->ressources->contains($ressource)) {
            $this->ressources->add($ressource);
            $ressource->addVoirRessource($this);
        }

        return $this;
    }

    public function removeRessource(Ressource $ressource): static
    {
        if ($this->ressources->removeElement($ressource)) {
            $ressource->removeVoirRessource($this);
        }

        return $this;
    }

}

File: mobile/ressources_re_mobile/lib/classes/Utilisateur.dart
// ignore_for_file: file_names

import 'package:ressources_re_mobile/classes/Commentaire.dart';
import 'package:ressources_re_mobile/classes/Progression.dart';
import 'package:ressources_re_mobile/classes/Ressource.dart';
import 'package:ressources_re_mobile/classes/VoirRessource.dart';

class Utilisateur {
  int? id;
  String? email;
  String? nom;
  String? prenom;
  List<String> roles = [];
  String? password;
  int? code;
  bool? verif;
  List<Ressource> proprietaireRessource = [];
  List<Commentaire> commentaires = [];
  List<Progression> progressions = [];
  List<VoirRessource> voirRessources = [];

  Utilisateur();

  int? getId() {
    return id;
  }

  String? getEmail() {
    return email;
  }

  void setEmail(String email) {
    this.email = email;
  }

  String? getNom() {
    return nom;
  }

  void setNom(String? nom) {
    this.nom = nom;
  }

  String? getPrenom() {
    return prenom;
  }

  void setPrenom(String? prenom) {
    this.prenom = prenom;
  }

  String getUserIdentifier() {
    return email ?? '';
  }

  List<String> getRoles() {
    List<String> roles = this.roles;
    roles.add('ROLE_USER');
    return roles.toSet().toList();
  }

  void setRoles(List<String> roles) {
    this.roles = roles;
  }

  String getPassword() {
    return password ?? '';
  }

  void setPassword(String password) {
    this.password = password;
  }

  int? getCode() {
    return code;
  }

  void setCode(int? code) {
    this.code = code;
  }

  bool? getVerif() {
    return verif;
  }

  void setVerif(bool? verif) {
    this.verif = verif;
  }

  void eraseCredentials() {}

  List<Ressource> getProprietaireRessource() {
    return proprietaireRessource;
  }

  void addProprietaireRessource(Ressource proprietaireRessource) {
    if (!this.proprietaireRessource.contains(proprietaireRessource)) {
      this.proprietaireRessource.add(proprietaireRessource);
      proprietaireRessource.setProprietaire(this);
    }
  }

  void removeProprietaireRessource(Ressource proprietaireRessource) {
    if (this.proprietaireRessource.remove(proprietaireRessource)) {
      if (proprietaireRessource.getProprietaire() == this) {
        proprietaireRessource.setProprietaire(null);
      }
    }
  }

  List<Commentaire> getCommentaires() {
    return commentaires;
  }

  void addCommentaire(Commentaire commentaire) {
    if (!this.commentaires.contains(commentaire)) {
      this.commentaires.add(commentaire);
      commentaire.setUtilisateur(this);
    }
  }

  void removeCommentaire(Commentaire commentaire) {
    if (this.commentaires.remove(commentaire)) {
      if (commentaire.getUtilisateur() == this) {
        commentaire.setUtilisateur(null);
      }
    }
  }

  List<Progression> getProgressions() {
    return progressions;
  }

  void addProgression(Progression progression) {
    if (!this.progressions.contains(progression)) {
      this.progressions.add(progression);
      progression.setUtilisateur(this);
    }
  }

  void removeProgression(Progression progression) {
    if (this.progressions.remove(progression)) {
      if (progression.getUtilisateur() == this) {
        progression.setUtilisateur(null);
      }
    }
  }

  List<VoirRessource> getVoirRessources() {
    return voirRessources;
  }

  void addVoirRessource(VoirRessource voirRessource) {
    if (!this.voirRessources.contains(voirRessource)) {
      this.voirRessources.add(voirRessource);
      voirRessource.setUtilisateur(this);
    }
  }

  void removeVoirRessource(VoirRessource voirRessource) {
    if (this.voirRessources.remove(voirRessource)) {
      if (voirRessource.getUtilisateur() == this) {
        voirRessource.setUtilisateur(null);
      }
    }
  }

  Utilisateur.fromJson(Map<String, dynamic> json)
  {
    id = json['id'];
    email = json['email'];
    nom = json['nom'];
    prenom = json['prenom'];
    roles = json['roles'] != null ? List<String>.from(json['roles']) : [];
    code = json['code'];
    verif = json['verif'];
  }

  
}
